src/test/kotlin/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/TestConnectionManager.kt

```diff
@@ -707,6 +709,190 @@ class TestConnectionManager {
+                        container,
+                        ConnectionManagerConfiguration(
+                            subscriberName = params.subscriberName,
+                            prefetchCount = DEFAULT_PREFETCH_COUNT,
+                            confirmationTimeout = DEFAULT_CONFIRMATION_TIMEOUT,
+                            minConnectionRecoveryTimeout = 100,
+                            maxConnectionRecoveryTimeout = 200,
+                            maxRecoveryAttempts = 5
+                        )
+                    )
+
+                    val consumerCounters: List<Counters> = List(params.consumers.size) { Counters() }
+
+                    class DeliverCallback(private val consumerNumber: Int) : ManualAckDeliveryCallback {
+                        override fun handle(
+                            deliveryMetadata: DeliveryMetadata,
+                            delivery: Delivery,
+                            confirmProcessed: ManualAckDeliveryCallback.Confirmation
+                        ) {
+                            val consumerCounter = consumerCounters[consumerNumber]
+
+                            LOGGER.info { "Received ${delivery.body.toString(Charsets.UTF_8)} from \"${delivery.envelope.routingKey}\"" }
+                            if (consumerCounter.messages.getAndIncrement() == 1) {
+                                LOGGER.info { "Left this message unacked" }
+                            } else {
+                                confirmProcessed.confirm()
+                                LOGGER.info { "Confirmed!" }
+                            }
+
+                            if (delivery.envelope.isRedeliver) {
+                                consumerCounter.redeliveredMessages.incrementAndGet()
+                            }
+                        }
+                    }
+
+                    val consumersThreads = params.consumers.mapIndexed { index, subscriberParams ->
+                        thread {
+                            val subscriberMonitor = connectionManager.basicConsume(params.queueName, DeliverCallback(index)) {
+                                LOGGER.info { "Canceled $it" }
+                            }
+
+                            Thread.sleep(1000)
+
+                            if (subscriberParams.unsubscribe) {
+                                LOGGER.info { "Unsubscribing..." }
+                                subscriberMonitor.unsubscribe()
+                            }
+                        }
+                    }
+
+                    repeat(params.messagesToSend) {
+                        LOGGER.info { "Sending message ${it + 1} to queue ${params.queueName}" }
+                        putMessageInQueue(container, params.queueName)
+                    }
+
+                    TestCaseContext(connectionManager, consumersThreads, consumerCounters)
+                }
+
+                LOGGER.info { "Sleeping..." }
+                Thread.sleep(63000)
```

Comment:
Could you please advise why 63 seconds exactly? Maybe we need to rethink the test structure here to avoid such long timeouts. Also, is there any guarantee that 63 seconds is enough?

Reply:
We need to wait 1 second for consumer timeout expiration. Then we need to wait until rabbitmq checks channels for timeout, it performs such check once a minute, so we need to wait for 60+ seconds. And then we need to wait for a while for recovering of connection and resubscribing channels. If 63 seconds will not be enough (e.g. in newer versions of rabbitmq) the test will fail, and we will fix it.

Reply:
Thank you for your explanation. I think that would be helpful for "future you" or me or anybody else if we leave a comment about that in the test. Because I don't believe it is possible to always remember this specific rabbitmq behavior and our internal logic